### PR TITLE
add solutions and editorials to usaco-646 and cses-1676

### PR DESCRIPTION
**Pull request by Amaryllidaceae**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [X] I have tested my code.
- [X] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/usaco-646.mdx [hunk 1]
```diff
@@ -2,21 +2,97 @@
...
 ## Implementation
 
+**Time Complexity:** $\mathcal{O}(V \cdot E \cdot \alpha(n)))$
+
 <LanguageSection>
 
+<PySection>
+
+```py
+//BeginCodeSnip{Disjoint Set Union}
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:33]: you can use python comments for code snip

File: solutions/cses-1676.mdx [hunk 2]
```diff
@@ -2,17 +2,25 @@
...
 ---
 
-**Time Complexity:** $\mathcal{O}(N\log N)$
+## Explanation
 
-<LanguageSection>
+We can represent the problem as an undirected graph, with each of the $n$ cities being a node and each of the $m$ roads being an edge. We then have two subproblems we need to consider:
 
-<CPPSection>
+Our first subproblem is counting the number of components each day. By the definition of a component given in the problem, a component is a connected component. We can count them using Disjoint Set Union (DSU), where each tree in the structure represents a connected component. We start off with $n$ cities and no roads between them, and thus $n$ connected components. For every road that is built, we union the two cities the road connects. If the union is successful, then two different connected components have been joined into one, and thus the number of connected components decreases by 1.
```
Comment (ChuangSheep) [inline solutions/cses-1676.mdx:12]: > Our first subproblem is counting the number of components each day. By the definition of a component given in the problem, a component is a connected component. 

idk maybe just `Our first subproblem is counting the number of connected components each day`?

File: solutions/usaco-646.mdx [hunk 3]
```diff
@@ -2,21 +2,97 @@
...
+author: Nathan Gong, Melody Yu, George Pong
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_closing_gold_open16.html)
 
+## Explanation
+
+We can represent the farm as an undirected graph, with each of the $n$ barns being a node, and each of the $m$ bidirectional paths being an edge. At each point in time, we want to check if the number of connected components is equal to one. Naively, this can be done with BFS or DFS. However, we would need to run this for all $n+1$ states of the barn, resulting in a TLE.
+
+We can use disjoint set union (DSU) to efficiently keep track of the number of connected components after each change. However, the key part of this solution is that, as DSU can only efficiently create relationships between nodes and not destroy them, we simulate closing the barn in reverse.
```
Comment (ChuangSheep) [inline solutions/usaco-646.mdx:14]:
```suggestion
We can use Disjoint Set Union (DSU) to efficiently keep track of the number of connected components after each change. However, the key part of this solution is that, as DSU can only efficiently create relationships between nodes and not destroy them, we simulate closing the barn in reverse.
```

**Review by ChuangSheep — COMMENTED** [review]
great!
<del>(just went through the explanations as I dont really review python</del>

File: solutions/cses-1676.mdx [hunk 4]
```diff
@@ -56,9 +64,61 @@ int main () {
 
 </CPPSection>
 
-<JavaSection>
+<PySection>
+
+```python
+import sys
+input = sys.stdin.readline
```
Comment (SansPapyrus683) [inline solutions/cses-1676.mdx:71]: what's this for

File: solutions/cses-1676.mdx [hunk 5]
```diff
@@ -56,9 +64,61 @@ int main () {
...
+
+# BeginCodeSnip{Disjoint Set Union}
```
Comment (SansPapyrus683) [inline solutions/cses-1676.mdx:73]: you can remove the spaces

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
yeah this code is good

File: solutions/cses-1676.mdx (hunk 4 above)
Comment (Amaryllidaceae) [reply solutions/cses-1676.mdx:71]: making input faster

Comment (Amaryllidaceae) [reply solutions/cses-1676.mdx:71]: redefining input makes TC#8 0.60 seconds while without it it's 1 second, barely passing

File: solutions/cses-1676.mdx (hunk 5 above)
Comment (Amaryllidaceae) [reply solutions/cses-1676.mdx:73]: just the one in the beginning? or all the spaces?

File: solutions/cses-1676.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/cses-1676.mdx:71]: can you add a comment saying this is the case

File: solutions/cses-1676.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/cses-1676.mdx:73]: all codesnip spaces

File: solutions/cses-1676.mdx [hunk 6]
```diff
@@ -2,17 +2,25 @@
...
 ---
 
-**Time Complexity:** $\mathcal{O}(N\log N)$
+## Explanation
 
-<LanguageSection>
+We can represent the problem as an undirected graph, with each of the $n$ cities being a node and each of the $m$ roads being an edge. We then have two subproblems we need to consider:
 
-<CPPSection>
+Our first subproblem is counting the number of connected components each day. We can count them using Disjoint Set Union (DSU), where each tree in the structure represents a connected component. We start off with $n$ cities and no roads between them, and thus $n$ connected components. For every road that is built, we union the two cities the road connects. If the union is successful, then two different connected components have been joined into one, and thus the number of connected components decreases by 1.
```
Comment (SansPapyrus683) [inline solutions/cses-1676.mdx:12]: i'd use unite instead of union as a verb

File: solutions/cses-1676.mdx [hunk 7]
```diff
@@ -2,17 +2,25 @@
...
+
+Our second subproblem is finding the size of the largest connected component each day. Conveniently, the DSU optimization union-by-size stores the size of any node's component in that tree's parent node. Furthermore, the largest component will only change if a union is successful, as otherwise the graph stays the same. If a union happens, we check if the size of the tree which was added to during union-ing is a new maximum.
```
Comment (SansPapyrus683) [inline solutions/cses-1676.mdx:14]: same here

File: solutions/cses-1676.mdx [hunk 8]
```diff
@@ -56,9 +64,61 @@ int main () {
...
+		self.size[a] += self.size[b]
+		return True
+
+	def connected(self, a: int, b: int) -> bool:
+		return self.find_parent(a) == self.find_parent(b)
+#EndCodeSnip
+
+n, m = map(int, input().split())
+
+dsu = DisjointSetUnion(n)
```
Comment (SansPapyrus683) [inline solutions/cses-1676.mdx:102]: i'd call this `cities` to represent the problem, but up to u

**Review by SansPapyrus683 — APPROVED** [review]
hell yeah

**Comment by Amaryllidaceae** [comment]
can someone review and merge?

**Comment by SansPapyrus683** [comment]
@ChuangSheep review

**Review by ChuangSheep — APPROVED** [review]
> <del>(just went through the explanations as I dont really review python</del>

emm ok, from what I see it all looks fine

**Comment by SansPapyrus683** [comment]
oh bit late but uh
could you show AC screens or submission links for these? just wanna make sure the solutions work

**Comment by Amaryllidaceae** [comment]
firefox_AQ1Gg9v2TU
firefox_WGSNT3TfoD

**Pull request merged**